### ISSUE TITLE
Extremely slow push performance when sending larger payload data with push (circa 200 per second)

### ISSUE THREAD

**Issue by neilmorton** [sideshow/apns2#163]
1. What version of Go are you using? `go version`
go version go1.13.1 linux/amd64

2. What OS and processor architecture are you using? `go env`
GOHOSTARCH="amd64"
GOHOSTOS="linux"

3. What did you do?
Using the Channel example https://github.com/sideshow/apns2/blob/master/_example/channel/main.go, I am trying to push upwards of 100,000 notifications, each with a payload of circa 1500 bytes (APNS limit is 4KB).

4. What did you expect to see?
Using 1 connection, and 1000 workers, I expected to see 1000 notifications being sent in parallel, multiplexed over the H2 connection, and for the workers to process all notifications. (APNS reporting 1000 Max Concurrent Streams).

5. What did you see instead?
I see 54 notifications sent initially, but then nothing is transferred until ACKs are received, at which time another 23 notifications are sent. Again nothing more is transferred until ACKs are received. This continues to repeat.

**Further Information**
APNS RTT is ~150ms (Server in London), so this takes a long time to process with a ~150ms delay after every 23 pushes.

Looking in WireShark, it appears that APNS are allowing TCP window scaling 7 (128), yet bytes in flight seem to stay below 64KB at all times.

If I open two connections (from the same server) and share the 1,000 workers (so 500 each, alternating), the speed is much quicker. If I create 3 connections, and spread the 1,000 workers over the 3 connections evenly then this is quicker again.

Of course in the multi connection scenarios the 1,000 stream limit on each connection is not fully utilised.

As mentioned above, it appears that the in-flight bytes are staying below 64KB at all times on each connection, despite APNS advertising a wscale of 7 (128). I see that the Receive Window starts at 28KB, but rapidly increases. According to Wireshark, the receive window on the APNS side is quickly growing to over 150KB, but looking at the in flight traffic, there appears to be a ceiling below 64KB that is being hit, with generally circa 40KB in flight.

It doesn't appear to be a CWND issue (usually around 308).

**Current Thoughts**
My feeling right now is that I am coming up against some sort of buffer limit, be that a default within my networking stack, or something in the GOLANG NET code perhaps? Alternatively, is there some sort of limit related to H2 PUSH (again, perhaps configurable?).

I have already reached out to @sideshow about this, who has confirmed that using multiple connections achieved much higher throughput. 

Whilst I can create multiple connections (I have created 40 without any issue, and this reduced the delivery time to acceptable levels), I would really like to get to the bottom of this, and feel that with the upcoming depreciation of the older APNS API, that many of us would like to get this sorted.

I should note that I do have a DTS ticket open with Apple around this, and whilst that hasn't uncovered anything of significant interest, I could use it in conjunction with any feedback here if needed.

**Wireshark graph**
Here is a Wireshark window scaling graph showing how the traffic is not approaching the receive window.
<img width="1424" alt="apns2_tcpdump_limited" src="https://user-images.githubusercontent.com/10706434/72525570-5efce480-385c-11ea-8b55-7fa333c5d877.png">

Thanks.

**Comment by sideshow** [comment]
Hi Neil, We've been using this to send over a billion notifications per month and haven't had any problems with speed.  You should be able to blast though 150,000 notifications in a few seconds, even on modest hardware which makes me think it's something about your setup. There definitely could be room for improvement in the transport/connection layer but given the fact that you are only able to send 200/s I am pretty sure your problem is higher up.

I wasn't aware you were trying to send from London and so latency from there is very likely a problem. You shouldn't need anything like 1000 workers to send that many notifications. Are you able to test this on AWS EC2 in us east west region and see if that resolves you issue.

Also, 

- What are your expectations in terms of timeframe for sending 150k messages 
- Can you share your example code?
- What is your internet connection speed? upload/download
- What are your machine specs? ie, Cores CPU, ram.
- How much of this have you allocated to the app? ie, if its running on docker etc
- Are you using the production apple gateway or is this in dev mode

**Comment by neilmorton** [comment]
Hi Adam, thanks for getting back to me.

Sending from London and the latency is certainly something. That said, I did spin up a Linode in Newark a few days ago which had a ~80ms response. Whilst faster (circa 2X), it was still a) relatively slow and b) not appearing to use the full receive window of the APNS. 

As you suggest, having the server closer to the APNS will certainly help due to the reduced latency, therefore less time waiting for ACKs and so more data can be sent, and with Newark this would result is a ~2X improvement). Point b though seems to be rather significant in that the connection is not being used to the max (APNS receive window). Looking at the receive window on the APNS and the fact that the in flight data is nowhere near it, seems to suggest that a near 12X performance improvement may be possible on the large payload notifications, and a 6X - 8X imprudent on the small payload notifications if the data was to be pushed out utilising much more of this window (see graphs below).

I should also note that I currently deliver these same notifications via the legacy API, albeit I appreciate there is no guarantee / acknowledgements with that API, and that delivers 100,000 notifications with the same payload in circa 50 seconds, hence my benchmark.

I have just re-run on my local server now with the ~1.5KB payload (with 998 workers to ensure it keeps the streams below 1,000 and thus another connection doesn't get opened up), and it has taken 45 seconds on 3 different runs.

I have also just swapped out the payload to a simple `payload.ContentAvailable()` and removed the extra payload data, and delivery time drops to within 4 seconds on 3 different runs.

Interestingly, on this run with no background data, the receive window / bytes out graph tells a somewhat different story in that APNS are raising the Window Size higher to ~650KB (which is nearly a flat line) and the bytes out have increased above where they are with the larger payload, often going above 80KB, and on occasions going above 100KB, albeit the connection still clearly oscillates based on the latency, and it is definitely not getting anywhere near the receive window.

The following is a Wireshark graph shows the profile with the small payload.
<img width="1425" alt="apns2_tcpdump_small" src="https://user-images.githubusercontent.com/10706434/72597487-cfac0b80-3905-11ea-88f7-fffcc227ca17.png">

The following is a Wireshark graph shows the profile with the larger payload (as before) but with 998 workers for direct comparison. This shows the Window Size being just under 500KB, but like in my previous post, the amount of Bytes Out in flight being generally under 40KB again. 
<img width="1418" alt="apns2_tcpdump_limited_2" src="https://user-images.githubusercontent.com/10706434/72599468-a7bea700-3909-11ea-8dc6-9de281212f33.png">
Of interest is that with 998 workers, 46 of the notifications get sent initially as opposed to 54, and then Framer gets into a pattern of writing ~46 DATA frames in chunks as ACKs are received.

Looking at the simple (small) payload, and comparing your figures for the slowest option [here](https://github.com/sideshow/apns2/wiki/APNS-HTTP-2-Push-Speed), then the delivery of over 2,500 per second (likely closer to 2,800 as it wasn't quite 4 full seconds) seems to be pretty good based on the latency, and not a million miles from your findings. I think from my findings with the Newark Linode would also tie in with this, in that being located on the East coast gave a ~2X latency improvement. However, the large payload is where I have the significant issue. 

_To answer your other questions:_
* Expectations in terms of timeframe: for sending 100k messages (with the larger payload) are that ideally they be sent within seconds 50 seconds (i.e. matching what I achieve with the existing legacy API), and not take over 7 minutes.
* Example code as follows:
```
package main

import (
	"database/sql"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"os"
	"runtime"

	_ "github.com/go-sql-driver/mysql"
	"github.com/sideshow/apns2"
	"github.com/sideshow/apns2/certificate"
	"github.com/sideshow/apns2/payload"
)

type Config struct {
	Database struct {
		Host     string `json:"host"`
		DBName   string `json:"dbname"`
		Username string `json:"username"`
		Password string `json:"password"`
	} `json:"database"`
	AuthKey struct {
		KeyPath string `json:"keypath"`
		KeyId   string `json:"keyid"`
		TeamId  string `json:"teamid"`
	} `json:"authkey"`
	Topic   string `json:"topic"`
	Logfile string `json:"logfile"`
	Logdir  string `json:"logdir"`
}

type Device struct {
	DeviceToken string `json:"devicetoken"`
	DeviceUUID  string `json:"uuid"`
}

func LoadConfiguration(filename string) (Config, error) {
	var config Config
	configFile, err := os.Open(filename)
	defer configFile.Close()
	if err != nil {
		return config, err
	}
	jsonParser := json.NewDecoder(configFile)
	err = jsonParser.Decode(&config)
	return config, err
}

func main() {

	config, _ := LoadConfiguration("push_config.json")

	environment := flag.String("environment", "", "Environment: dev / prod (REQUIRED)")
	location := flag.String("location", "", "Location: 0 / 1 (REQUIRED)")
	alert := flag.String("alert", "", "Alert level: yellow / amber / red / info_high / info_low / bgdataonly (REQUIRED)")
	message := flag.String("message", "", "Message text")
	activity := flag.String("activity", "", "activity data string (optional)")
	creationTime := flag.String("creationTime", "", "creationTime string (optional)")
	siteId := flag.String("siteId", "", "siteId string (optional)")
	thresholds := flag.String("thresholds", "", "thresholds string (optional)")
	maxage := flag.String("maxage", "", "maxage string (optional)")
	pushTime := flag.String("pushTime", "", "pushTime string (optional)")
	flag.Parse()

	os.MkdirAll(config.Logdir, 0755)
	config.Logfile = config.Logdir + "/" + config.Logfile + *environment + ".log"
	f, err := os.OpenFile(config.Logfile, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()
	log.SetOutput(f)

	sound := ""

	location := "uk"

	cert, err := certificate.FromP12File("./certs/prod.p12", "")
	if err != nil {
		log.Fatal("Cert Error:", err)
	}

	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@/%s", config.Database.Username, config.Database.Password, config.Database.DBName))
	defer db.Close()

	if err != nil {
		log.Fatal(err.Error())
	}

	err = db.Ping()
	if err != nil {
		log.Fatal(err.Error())
	}

	log.Println("Query Database...")
	//rows, err := db.Query("SELECT devicetoken, uuid FROM pushdevices WHERE location = ? AND environment = ? AND alert <= ? AND alert <> 0", location, *environment, alertNum)
	rows, err := db.Query("SELECT devicetoken, uuid FROM pushdevices WHERE location = ? AND environment = ? AND alert <= ? AND alert <> 0 limit 10000", location, *environment, alertNum)

	if err != nil {
		log.Fatal(err.Error())
	}

	var devices []Device

	for rows.Next() {
		var device Device
		if err := rows.Scan(&device.DeviceToken, &device.DeviceUUID); err != nil {
			log.Fatal(err.Error())
		}
		devices = append(devices, device)
	}

	if err := rows.Err(); err != nil {
		log.Fatal(err.Error())
	}

	// Setup notifications
	notifications := make(chan *apns2.Notification, len(devices))
	responses := make(chan *apns2.Response, len(devices))

	cores := runtime.NumCPU()
	runtime.GOMAXPROCS(cores)

	cores = 1 // limit temp for testing / switch up to get more connections

	clients := make(chan *apns2.Client, 100) // High enough to allow for multiple cores

	for i := 0; i < cores; i++ {
		if *environment == "prod" {
			clients <- apns2.NewClient(cert).Production()
		} else {
			// default to development if production is not selected
			clients <- apns2.NewClient(cert).Development()
		}
	}

	workers := 998

	for i := 0; i < workers; i++ {
		client := <-clients // grab a client from the pool
		clients <- client // add the client back to the pool
		go worker(client, notifications, responses, i)
	}

	log.Printf("Created %v workers, with %v clients.", workers, cores)

	payload := payload.NewPayload()
	if *alert != "bgdataonly" {
		payload.Alert(*message).ContentAvailable().Sound(sound).Custom("activity", activity).Custom("creationtime", creationTime).Custom("siteid", siteId).Custom("thresholds", thresholds).Custom("maxage", maxage).Custom("location", Location).Custom("pushtime", pushTime)
	} else {
		// just deliver updated data in the background (no user visible notification)
		payload.ContentAvailable().Custom("activity", activity).Custom("creationtime", creationTime).Custom("siteid", siteId).Custom("thresholds", thresholds).Custom("maxage", maxage).Custom("location", Location).Custom("pushtime", pushTime)
		//payload.ContentAvailable()
	}

	// And send
	for i := 0; i < len(devices); i++ {
		d := devices[i]
		if *alert != "bgdataonly" {
			// Priority set to High for Alert
			// PushType set to Alert
			n := &apns2.Notification{
				DeviceToken: d.DeviceToken,
				ApnsID:      d.DeviceUUID,
				Topic:       config.Topic,
				Payload:     payload,
				Priority:    apns2.PriorityHigh,
				PushType:    apns2.PushTypeAlert,
			}
			notifications <- n
		} else {
			// Priority MUST be set to Low when no alert / sound / badge (i.e. background only)
			// PushType should be set to Background when not delivering an alert
			n := &apns2.Notification{
				DeviceToken: d.DeviceToken,
				ApnsID:      d.DeviceUUID,
				Topic:       config.Topic,
				Payload:     payload,
				Priority:    apns2.PriorityLow,
				PushType:    apns2.PushTypeBackground,
			}
			notifications <- n
		}
	}

	for i := 0; i < len(devices); i++ {
		res := <-responses
		if res.StatusCode != 200 {
		// Code to delete token from DB if appropriate
		}
	}

	close(notifications)
	close(responses)
	close(clients)
}

func worker(client *apns2.Client, notifications <-chan *apns2.Notification, responses chan<- *apns2.Response, wid int) {
	for n := range notifications {
		res, err := client.Push(n)
		if err != nil {
			log.Fatal("Push Error:", err)
		}
		responses <- res
	}
}
```
* Connection speed 762.79 Mbps Up / 891.95 Mbps Down, just tested now.
* Machine Specs: Linode 2GB instance, 1 CPU Core, 2GB RAM.
* Not using Docker, so all of this is available.
* Using Production Apple Gateway.

In summary, I feel there is something blocking the ability to fully utilise the Receive Window advertised by APNS, and as a result, much of the time is just waiting for ACKs and not doing anything. This could be some setting on my server (although when I set up the Newark Linode that experienced the same issues), or it could be something with the management of TCP / H2 connections in GOLANG / my code / apns2, perhaps in that a I need to add a buffer setting or such like, and this is where it would be amazing if someone could help.

Thanks again.

**Comment by lishuhao** [comment]
@sideshow
> I wasn't aware you were trying to send from London and so latency from there is very likely a problem 

Yes it is .
600 notice/sec from China
2500 notice/sec from AWS EC2 in us east / west region

**Comment by neilmorton** [comment]
@lishuhao 

> 600 notice/sec from China
> 2500 notice/sec from AWS EC2 in us east / west region

That is interesting to hear the results you are seeing. 

As I mentioned, 

> Sending from London and the latency is certainly something. That said, I did spin up a Linode in Newark a few days ago which had a ~80ms response. Whilst faster (circa 2X), it was still a) relatively slow and b) not appearing to use the full receive window of the APNS.

As that mentioned, I saw just a 2X improvement. So, that is a large difference, but the difference relating to the payload size seems to have a lot to do with it.

It looks like I will end up placing a server in the US AND spawning multiple connections in order to get the larger payload data through in a timely manner.

**Comment by neilmorton** [comment]
In the end I implemented this using multiple clients which gets the throughput I need. 

I am definitely seeing what seems to be a limit to the amount of data that can be 'in flight' on any given connection of 64KB, despite the scale increasing to accommodate higher than this, which leads me to believe there is some inherent limit *somewhere*, whether on the APNS, the H2 library / buffer.

Testing with differing payload sizes appears to show that the number of 'in flight' notifications is linked to the amount of 'in flight' data that can exist (64KB) on any given connection. Therefore many more simple background push notifications could be sent in that 64KB window than notifications carrying larger payloads.

Increasing the number of workers above the number that creates enough notifications to fill this apparent 64KB 'in flight' window doesn't seem to provide any benefit, although I am allowing 2 X the number of possible in flight notifications as workers, in order to ensure back pressure.

For reference I tried a Linode in Fremont, CA, and saw very similar times to that located in London. Interestingly though, the path to Apple servers showed a hop via the UK (odd). I may have another go with a US based server when I get more time.

I shall close this now, but wanted to include this summary for future reference.

**Comment by ColeTownsend** [comment]
@neilmorton did you ever achieve your goal of 100,000 in under 50s?

**Comment by neilmorton** [comment]
@ColeTownsend just about! Looking at my latest logs, I am pushing 100,000 in just under 51 seconds.

I have it configured with 20 clients, coupled with 1000 workers, which appears to give best throughput for the full payload sent with each notification. This is based on a possible 23 or so notifications per window, so allowing 50 notifications per connection which keeps some back pressure. 

It is worth noting that I am sending a data payload with these notifications (not just a notification message) which adds an extra ~1300 bytes per notification.

That is to say that if I was not sending the extra data payload, I expect I would see a significant increase in throughput as there would be many more notifications fitting within the 64KB ceiling I mentioned (in my case possibly a 5 or 6 fold increase?).

**Comment by ColeTownsend** [comment]
thanks @neilmorton was curious what the settings were

**Comment by neilmorton** [comment]
@ColeTownsend - Sorry, only just seen this. 20 clients, with 1000 workers seemed to allow a manageable pressure on the connections. But I found it depended on size of notifications (mine have a large payload). 

It does depend on the server capability too.

What specifically setting wise?
